mitmproxy/proxy/layers/http/_http_h2.py

```diff
@@ -77,8 +79,16 @@ def send_data(
                 # We can't send right now, so we buffer.
                 self.stream_buffers[stream_id].append(SendH2Data(data, end_stream))
 
+    def send_trailers(self, stream_id, trailers):
+        if self.stream_buffers.get(stream_id, None):
+            # Though trailers are not subject to flow control, we need to queue them and send strictly after data frames
+            self.stream_trailers[stream_id] = trailers
+        else:
+            self.send_headers(stream_id, trailers, end_stream=True)
+
     def end_stream(self, stream_id) -> None:
-        self.send_data(stream_id, b"", end_stream=True)
+        if not self.stream_trailers.get(stream_id):
+            self.send_data(stream_id, b"", end_stream=True)
```

Comment:
This path here wasn't immediately clear to me, let's add a short explainer.
```suggestion
        if stream_id in self.stream_trailers:
            return  # we already have trailers queued up that will end the stream.
        self.send_data(stream_id, b"", end_stream=True)
```